File: LektaAndroid/app/src/main/java/com/lekta/app/services/MercadoPagoService.kt
package com.lekta.app.services

import com.lekta.app.Config
import com.lekta.app.models.CartItem
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.net.HttpURLConnection
import java.net.URL
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.math.floor

data class PreferenceResponse(
    val id: String,
    val initPoint: String,
    val sandboxInitPoint: String
)

enum class PaymentStatus { PENDING, APPROVED, REJECTED, CANCELLED }

@Singleton
class MercadoPagoService @Inject constructor(
    private val securePrefs: SecurePrefs
) {
    private val accessToken: String get() = securePrefs.mpAccessToken ?: ""

    companion object {
        private const val CONNECT_TIMEOUT_MS = 15_000
        private const val READ_TIMEOUT_MS = 30_000
    }

    private fun HttpURLConnection.applyDefaults() {
        connectTimeout = CONNECT_TIMEOUT_MS
        readTimeout = READ_TIMEOUT_MS
        useCaches = false
    }

    suspend fun createPreference(items: List<CartItem>): PreferenceResponse = withContext(Dispatchers.IO) {
        val conn = (URL("${Config.MP_BASE_URL}/checkout/preferences").openConnection() as HttpURLConnection).apply {
            applyDefaults()
            requestMethod = "POST"
            setRequestProperty("Authorization", "Bearer $accessToken")
            setRequestProperty("Content-Type", "application/json")
            doOutput = true
        }

        try {
            val mpItems = JSONArray()
            items.forEach { item ->
                mpItems.put(JSONObject().apply {
                    put("title", item.product.name)
                    put("quantity", item.quantity)
                    put("unit_price", item.product.finalPrice)
                    put("currency_id", "ARS")
                })
            }

            val total = items.sumOf { it.product.finalPrice * it.quantity }
            val marketplaceFee = floor(total * Config.MARKETPLACE_FEE_PERCENT / 100.0 * 100) / 100

            val body = JSONObject().apply {
                put("items", mpItems)
                put("marketplace_fee", marketplaceFee)
                put("marketplace", Config.MP_CLIENT_ID)
            }

            conn.outputStream.use { it.write(body.toString().toByteArray()) }

            if (conn.responseCode != 201) throw Exception("HTTP ${conn.responseCode}")

            val response = JSONObject(conn.inputStream.bufferedReader().use { it.readText() })
            PreferenceResponse(
                id = response.getString("id"),
                initPoint = response.getString("init_point"),
                sandboxInitPoint = response.getString("sandbox_init_point")
            )
        } finally {
            conn.disconnect()
        }
    }

    suspend fun checkPaymentStatus(preferenceId: String): PaymentStatus = withContext(Dispatchers.IO) {
        if (!preferenceId.all { it.isLetterOrDigit() || it == '-' || it == '_' }) {
            throw IllegalArgumentException("Invalid preference ID")
        }

        val conn = (URL("${Config.MP_BASE_URL}/checkout/preferences/$preferenceId").openConnection() as HttpURLConnection).apply {
            applyDefaults()
            setRequestProperty("Authorization", "Bearer $accessToken")
        }

        try {
            if (conn.responseCode !in 200..299) throw Exception("HTTP ${conn.responseCode}")

            val json = JSONObject(conn.inputStream.bufferedReader().use { it.readText() })
            val status = json.optString("status", "pending")
            PaymentStatus.entries.find { it.name.equals(status, ignoreCase = true) } ?: PaymentStatus.PENDING
        } finally {
            conn.disconnect()
        }
    }

    suspend fun verificarPOS(): Unit = withContext(Dispatchers.IO) {
        val conn = (URL("${Config.MP_BASE_URL}/pos").openConnection() as HttpURLConnection).apply {
            applyDefaults()
            setRequestProperty("Authorization", "Bearer $accessToken")
        }
        try {
            if (conn.responseCode !in 200..299) throw Exception("HTTP ${conn.responseCode}")
        } finally {
            conn.disconnect()
        }
    }

    suspend fun obtenerTotalPagos(desde: Date): Double = withContext(Dispatchers.IO) {
        val fmt = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'", Locale.US).apply {
            timeZone = TimeZone.getTimeZone("UTC")
        }

        val conn = (URL(
            "${Config.MP_BASE_URL}/v1/payments/search" +
                    "?status=approved" +
                    "&begin_date=${fmt.format(desde)}" +
                    "&end_date=${fmt.format(Date())}" +
                    "&limit=100"
        ).openConnection() as HttpURLConnection).apply {
            applyDefaults()
            setRequestProperty("Authorization", "Bearer $accessToken")
        }

        try {
            if (conn.responseCode !in 200..299) throw Exception("HTTP ${conn.responseCode}")

            val json = JSONObject(conn.inputStream.bufferedReader().use { it.readText() })
            val results = json.getJSONArray("results")
            var total = 0.0
            for (i in 0 until results.length()) {
                total += results.getJSONObject(i).getDouble("transaction_amount")
            }
            total
        } finally {
            conn.disconnect()
        }
    }
}

File: Services/MercadoPagoService.swift
import Foundation

struct PreferenceResponse: Codable {
    let id: String
    let initPoint: String
    let sandboxInitPoint: String

    enum CodingKeys: String, CodingKey {
        case id
        case initPoint = "init_point"
        case sandboxInitPoint = "sandbox_init_point"
    }
}

enum PaymentStatus: String, Codable {
    case pending, approved, rejected, cancelled
}

struct MercadoPagoService {
    static var accessToken: String {
        KeychainService.mpAccessToken ?? ""
    }

    // MARK: - Create preference

    static func createPreference(items: [CartItem]) async throws -> PreferenceResponse {
        let url = URL(string: "https://api.mercadopago.com/checkout/preferences")!
        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")

        let mpItems = items.map { item -> [String: Any] in
            [
                "title": item.product.name,
                "quantity": item.quantity,
                "unit_price": item.product.finalPrice,
                "currency_id": "ARS"
            ]
        }
        let body: [String: Any] = ["items": mpItems]
        request.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, http.statusCode == 201 else {
            throw URLError(.badServerResponse)
        }
        return try JSONDecoder().decode(PreferenceResponse.self, from: data)
    }

    // MARK: - Check payment status

    static func checkPaymentStatus(preferenceId: String) async throws -> PaymentStatus {
        let url = URL(string: "https://api.mercadopago.com/checkout/preferences/\(preferenceId)")!
        var request = URLRequest(url: url)
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")

        let (data, _) = try await URLSession.shared.data(for: request)

        struct StatusResponse: Codable {
            let status: String?
        }
        let decoded = try JSONDecoder().decode(StatusResponse.self, from: data)
        return PaymentStatus(rawValue: decoded.status ?? "pending") ?? .pending
    }

    // MARK: - Caja: verificar POS activo

    /// Verifica que el POS configurado en MP esté activo.
    /// Endpoint: GET /pos  — devuelve la lista de puntos de venta del vendedor.
    static func verificarPOS() async throws {
        let url = URL(string: "https://api.mercadopago.com/pos")!
        var request = URLRequest(url: url)
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")

        let (_, response) = try await URLSession.shared.data(for: request)
        guard let http = response as? HTTPURLResponse, (200...299).contains(http.statusCode) else {
            throw URLError(.badServerResponse)
        }
    }

    // MARK: - Caja: total de pagos aprobados desde una fecha

    /// Suma todos los pagos aprobados vía MP desde `desde` hasta ahora.
    /// Endpoint: GET /v1/payments/search
    static func obtenerTotalPagos(desde: Date) async throws -> Double {
        let formatter = ISO8601DateFormatter()
        let begin = formatter.string(from: desde)
        let end   = formatter.string(from: Date())

        var components = URLComponents(string: "https://api.mercadopago.com/v1/payments/search")!
        components.queryItems = [
            .init(name: "status",     value: "approved"),
            .init(name: "begin_date", value: begin),
            .init(name: "end_date",   value: end),
            .init(name: "limit",      value: "100")
        ]
        var request = URLRequest(url: components.url!)
        request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")

        let (data, _) = try await URLSession.shared.data(for: request)

        struct SearchResult: Codable {
            struct Payment: Codable {
                let transactionAmount: Double
                enum CodingKeys: String, CodingKey {
                    case transactionAmount = "transaction_amount"
                }
            }
            struct Paging: Codable { let total: Int }
            let results: [Payment]
        }

        let decoded = try JSONDecoder().decode(SearchResult.self, from: data)
        return decoded.results.reduce(0) { $0 + $1.transactionAmount }
    }
}
